staging/src/k8s.io/kube-aggregator/pkg/apiserver/openapi_aggregator.go

```diff
@@ -250,21 +322,74 @@ func (s *openAPIAggregator) loadApiServiceSpec(handler http.Handler, apiService
 		return nil
 	}
 
-	openApiSpec, err := s.downloadOpenAPISpec(handler)
+	openApiSpec, etag, _, err := s.downloadOpenAPISpec(handler, "")
 	if err != nil {
 		return err
 	}
+	if openApiSpec == nil {
+		// API service does not provide an OpenAPI spec, ignore it.
+		return nil
+	}
 	aggregator.FilterSpecByPaths(openApiSpec, []string{"/apis/" + apiService.Spec.Group + "/"})
 
+	_, existingService := s.openAPISpecs[apiService.Name]
 	s.openAPISpecs[apiService.Name] = &openAPISpecInfo{
 		apiService: *apiService,
 		spec:       openApiSpec,
+		handler:    handler,
+		etag:       etag,
 	}
 
 	err = s.updateOpenAPISpec()
 	if err != nil {
 		delete(s.openAPISpecs, apiService.Name)
 		return err
 	}
+	if !existingService {
+		s.openAPIAggregationController.enqueue(apiService.Name)
+	}
+
 	return nil
 }
+
+func (s *openAPIAggregator) removeApiServiceSpec(apiServiceName string) error {
+	if _, existingService := s.openAPISpecs[apiServiceName]; existingService {
+		oldSpecs := s.openAPISpecs
+		delete(s.openAPISpecs, apiServiceName)
+		err := s.updateOpenAPISpec()
+		if err != nil {
+			s.openAPISpecs = oldSpecs
+			return err
+		}
+	}
+	return nil
+}
+
+func (s *openAPIAggregator) UpdateApiServiceSpec(apiServiceName string) (shouldBeRateLimited, deleted bool, err error) {
```

Comment:
this func is not thread save. This should be at least documented, especially as it called from the controller.